### ISSUE TITLE
Synchronized block deadlock  after an exceptions on iOS

### ISSUE THREAD

**Issue by PiotrZub** [codenameone/CodenameOne#1519]
I prepare a testcase to show hangup after exception.
https://drive.google.com/file/d/0B_J2xdfDrOC8LWg4Sk1TV2FJZmc/view?usp=sharing

Full problem discussion is on forum:
https://groups.google.com/forum/embed/?place=forum%2Fcodenameone-discussions&showsearch=true&showpopout=true&hl=en&parenturl=https%3A%2F%2Fcodenameone.com%2Fdiscussion-forum.html&afterlogin=#!searchin/codenameone-discussions/synchronized/codenameone-discussions/XImndrLkZ9E/NyK-YHuxMSIJ
The app is hangup on iOS and working with out any problem on simulator.
At the beginning it test how app is working with java exception than it test with aritmetic exception.
Result from emulator:
https://docs.google.com/document/d/11b6SilNydj4tN8ZU1pq5w8tPsq8prviCu9g2ur3I4ss/edit?usp=sharing

Ps. At the beging I get:
java.lang.NullPointerException
    at com_codename1_io_Log.p:241
    at com_codename1_io_Log.p:231
    at eu_forann_fmapapp_LogF.e:68
    at eu_forann_fmapapp_LogF.p:78
    at eu_forann_fmapapp_TestClass_2.run:46
    at eu_forann_fmapapp_ThreadCatch.runCatch:71
    at eu_forann_fmapapp_ThreadCatch.run:47
it shouldn't happen because string isn;t null, may be it is another bug?

**Comment by codenameone** [comment]
This test case is a bit noisy and hard to follow. Please try using cleaner handcoded test cases in the future, e.g. for this issue this test case is much easier to work with:

    public void start() {
        if(current != null){
            current.show();
            return;
        }
        final Form hi = new Form("Issue 1425");
        hi.setLayout(new BoxLayout(BoxLayout.Y_AXIS));
        Button button = new Button("Test");
        hi.addComponent(button);
        button.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent evt) {
                for(int iter = 0 ; iter < 10 ; iter++) {
                    final int currentThread = iter;
                    new Thread() {
                        @Override
                        public void run() {
                            Object o = null;
                            try {
                                synchronized(LOCK) {
                                    o.equals(null);
                                }
                            } catch(Exception err) {
                                Display.getInstance().callSerially(new Runnable() {
                                    public void run() {
                                        hi.addComponent(new Label("Caught " + currentThread));
                                        hi.getContentPane().revalidate();
                                    }
                                });
                            }
                        }
                    }.start();
                }
            }
        });
        hi.show();
    }

**Comment by PiotrZub** [comment]
Ok. I think that my noisy test could be used in second step because there is more change to get some unexpected situation.
Is there any change to solve the issue now? 
I think it it serious problem for all CN1 user even for enterprise because an app can be blocked randomly so it shouldn't be used as ready to use.

**Comment by codenameone** [comment]
Its a bit complex, we are looking at this.

**Comment by PiotrZub** [comment]
Of course, sorry but there wasn't any status of this issue that is why I was worried you forgot.

**Comment by codenameone** [comment]
This seems to have caused some regressions... https://groups.google.com/d/msg/codenameone-discussions/wB3qD4ulSyw/sKiGHftNo_QJ A bit hard to nail down a test case though.

**Comment by codenameone** [comment]
Hopefully these regressions are now fixed, I made a change that's a bit risky for #1540 which should keep this fix still working.

**Comment by codenameone** [comment]
This is now finally fixed. It was a deep long series of issues the last of which is now finally online. It seems that when two try blocks started together they were generated in the wrong order.
